Stack unification and test

mprse opened the pull request:
### Description

This PR provides test which verifies stack sizes:
- ISR stack: exactly 1 kB,
- Main thread stack:  exactly 4 kB,
- Default thread stack: exactly 4 kB,
- RTOS-less thread stack: at least 4 kB - skipped for now since it looks like it is impossible to check RTOS-less thread (tests are built with RTOS).

Additionally PR provides fixes in linker scripts for some example targets. Can someone who has more experience with linker scripts check if these changes are correct? I could then proceed with the further targets.

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

mprse committed:
Stack size unification test.

Test checks stack sizes:
 - ISR stack exactly 1kB
 - Main thread stack, exactly 4kB
 - Thread stack, exactly 4kB
 - RTOS-less thread stack, at least 4kB - skipped for now since it looks like it is impossible to check RTOS-less thread (tests are built with RTOS).

bulislaw commented:
@SeppoTakalo fyi

SeppoTakalo reviewed (dismissed):
Looks good.

SeppoTakalo commented:
Can you also add checks for the boot time stack?

I believe those are the same as RTOS-less builds. For example IAR seems to define it as `define symbol __ICFEDIT_size_cstack__   = 0x400;`
and it is not the same as main thread stack, as main thread stack is allocated here: https://github.com/ARMmbed/mbed-os/blob/master/rtos/TARGET_CORTEX/mbed_boot.c#L189

So if we set stack size in linker file to 4kB it is used in boot, before RTX starts, and then we allocate another 4kB for main thread and kick the kernel to run.

However, RTX seems to suggest that 300 bytes should be enough: https://www.keil.com/pack/doc/CMSIS/RTOS/html/theory.html

SeppoTakalo commented:
Or here: http://www.keil.com/support/man/docs/rlarm/rlarm_ar_technical_data.htm
it tells that 300 + 128 bytes for kernel.
which to me suggest, that within a linker file 428 should be the stack size we use. Not 4 kB

mprse commented:
@SeppoTakalo In the test I am checking `mbed_stack_isr_size` which is defined based on `CSTACK`, `ISR_STACK_SIZE` symbols:
https://github.com/ARMmbed/mbed-os/blob/eb3d3fdb54df3de967b93abff6259e2eb70fd42b/rtos/TARGET_CORTEX/mbed_boot.c#L225-L232
https://github.com/ARMmbed/mbed-os/blob/19c795c31202d0bfb3b0f04bcdec29d20c9b3943/rtos/TARGET_CORTEX/mbed_boot.c#L266-L275

For example STM boards defines `ISR_STACK_SIZE` as follows (`GCC_ARM`):
https://github.com/ARMmbed/mbed-os/blob/eb3d3fdb54df3de967b93abff6259e2eb70fd42b/targets/TARGET_STM/mbed_rtx.h#L136

Now for specific `TARGET_STM32L476xG` family we have the following definitions:
`IAR`
https://github.com/ARMmbed/mbed-os/blob/eb3d3fdb54df3de967b93abff6259e2eb70fd42b/targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L476xG/device/TOOLCHAIN_IAR/stm32l476xx.icf#L25-L28
`GCC_ARM`
https://github.com/ARMmbed/mbed-os/blob/eb3d3fdb54df3de967b93abff6259e2eb70fd42b/targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L476xG/device/TOOLCHAIN_GCC_ARM/STM32L476XX.ld#L158-L165
`ARM`
Not defined

According to the requirements I'm going to modify ISR stack for all targets to 1 kB. At least for `GCC` and `IAR`. I can't see how ISR stack size is defined for `ARM` compiler.
As far as I understand you suggest to set ISR stack size to ~0,5 kB?

deepikabhavnani commented:
> According to the requirements I'm going to modify ISR stack for all targets to 1 kB

Why set it to 1KB only? Some devices might need higher ISR stack can we have some good range instead?

deepikabhavnani commented:
> ARM Not defined

Below files show where ISR stack is set for ARM. Look for what is set in ISR_STACK_START for all targets. Many of them use same define "ARM_LIB_STACK" in scatter files

https://github.com/ARMmbed/mbed-os/blob/7b428916f5b35da28a1b18e2c149c1115bb89d92/targets/TARGET_NUVOTON/mbed_rtx.h#L31

https://github.com/ARMmbed/mbed-os/blob/097720670eb1c569b19af85fefcf61f202c39aa3/targets/TARGET_NUVOTON/TARGET_NANO100/device/TOOLCHAIN_ARM_STD/NANO130.sct#L13

SeppoTakalo commented:
@mprse 
> As far as I understand you suggest to set ISR stack size to ~0,5 kB?

No, this was based on my original understanding that boot time stack/main stack and ISR stacks are different.

But according to [Cortex-M3 User Guide Reference Material: 2.1.2. Stacks](http://infocenter.arm.com/help/topic/com.arm.doc.dui0552a/BABDGADF.html)
`In Thread mode, the CONTROL register controls whether the processor uses the main stack or the process stack, see CONTROL register. In Handler mode, the processor always uses the main stack. `

So clearly "boot" stack and ISR stacks are the same. And I tried to find, where we set that pointer, but it seems to be the first vector, so its the linker file + `startup_*.S`

According the [Cortex-M3 Devices Generic User Guide: Stack Pointer](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0552a/CHDBIBGJ.html)
```
The Stack Pointer (SP) is register R13. In Thread mode, bit[1] of the CONTROL register indicates the stack pointer to use:
0 = Main Stack Pointer (MSP). This is the reset value.
1 = Process Stack Pointer (PSP).
On reset, the processor loads the MSP with the value from address 0x00000000.
```

So I was wrong assuming that checking ISR stack would not be enough, but those are the same.

Should we actually change the name, because my understanding is now that both RTX kernel and ISR both use this "main stack" where `MSP` register points, and each thread have their own stack where `PSP` points?

And @deepikabhavnani why would some devices need 4kB of ISR stack? Sound like a lot compared to RTX kernel requiring only 400 bytes. Do you know some examples?

At least within Mbed OS it is very limited what you can do in ISR, so the call path would be usually very short.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52832/device/TOOLCHAIN_IAR/nRF52832.icf:
```diff
@@ -40,7 +40,7 @@ export symbol __ICFEDIT_region_RAM_end__;
 
 /*-Sizes-*/
 /*Heap 1/4 of ram and stack 1/8*/
-define symbol __ICFEDIT_size_cstack__   = 0x800;
```
That would break BLE.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF51/TARGET_MCU_NRF51822_UNIFIED/device/TOOLCHAIN_IAR/TARGET_MCU_NORDIC_32K/nRF51822_QFAA.icf:
```diff
@@ -11,8 +11,8 @@ define symbol __ICFEDIT_region_RAM_end__   = 0x20007FFF;
 export symbol __ICFEDIT_region_RAM_end__;
 /*-Sizes-*/
-/*Heap 1/4 of ram and stack 1/8*/
-define symbol __ICFEDIT_size_cstack__   = 0x800;
```
That would break BLE.

pan- reviewed (changes requested):
IRQ Stack must be at least 2K large on Nordic targets. This is not an option to change it.

deepikabhavnani commented:
> And @deepikabhavnani why would some devices need 4kB of ISR stack

I never meant 4KB stack size as requirement. I would suggest to have a range based on analysis / input from target devices instead of hard-code 1KB.

mprse commented:
Thanks for all your comments. They are really helpful!

> Should we actually change the name, because my understanding is now that both RTX kernel and ISR both use this "main stack" where MSP register points, and each thread have their own stack where PSP points?

I agree that the name ("Interrupt stack") is misleading and it would be better to be consistent with the documentation and use names "main stack" (interrupt and boot time stack) and "process stack" (threads).

> I never meant 4KB stack size as requirement. I would suggest to have a range based on analysis / input from target devices instead of hard-code 1KB.

I also agree that setting main stack/isr stack size to 1 KB for all boards might not be a good solution. For many cases it should be ok, but for sure there are cases when it is impossible (like BLE support on NRF boards pointed by @pan).

I found the following problematic cases:

- Renesas targets are based on Cortex-A and stack is organised in the different way. We have different exception modes like Undefined instruction, Supervisor call, Abort, IRQ,  FIQ and for each of them separate stack is used. Total stack size is defined as a sum of all these stacks which gives 92 KB.

https://github.com/ARMmbed/mbed-os/blob/443e37b00c6882756bdeecfd41f45cdd2efaf872/targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_RZ_A1H/device/TOOLCHAIN_ARM_STD/MBRZA1H.sct#L44

https://github.com/ARMmbed/mbed-os/blob/443e37b00c6882756bdeecfd41f45cdd2efaf872/targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_RZ_A1H/device/TOOLCHAIN_ARM_STD/mem_RZ_A1H.h#L58-L63

https://github.com/ARMmbed/mbed-os/blob/443e37b00c6882756bdeecfd41f45cdd2efaf872/targets/TARGET_RENESAS/TARGET_RZ_A1XX/TARGET_RZ_A1H/device/TOOLCHAIN_ARM_STD/startup_RZ_A1H.c#L116-L139

- On some boards with very small amount of RAM the main stack/isr stack size is currently less than 1 KB.  
`TARGET_KL05Z`: main stack/isr stack size is set to 512 B.  This board has only 4 KB RAM.

https://github.com/ARMmbed/mbed-os/blob/443e37b00c6882756bdeecfd41f45cdd2efaf872/targets/TARGET_Freescale/TARGET_KLXX/TARGET_KL05Z/device/TOOLCHAIN_IAR/MKL05Z4.icf#L14-L17

`TARGET_NCS36510`: main stack/isr stack size is set to 512 B.  This board has 48 KB RAM.

https://github.com/ARMmbed/mbed-os/blob/443e37b00c6882756bdeecfd41f45cdd2efaf872/targets/TARGET_ONSEMI/TARGET_NCS36510/device/TOOLCHAIN_IAR/NCS36510.icf#L31-L33

`TARGET_LPC810`: main stack/isr stack size is set to 256 B. This  micro-controller has only 4 KB RAM.

https://github.com/ARMmbed/mbed-os/blob/443e37b00c6882756bdeecfd41f45cdd2efaf872/targets/TARGET_NXP/TARGET_LPC81X/TARGET_LPC810/device/TOOLCHAIN_IAR/LPC810.icf#L13-L15

I'm not sure if increasing main stack/isr stack size for boards which have only 4KB or 8KB RAM is possible.

@bulislaw I think we should consider changing the requirement. I still did not checked all targets, but it looks like I can not simply unify main stack/isr stack size to 1KB for all targets. At least some exceptions are needed.

bulislaw commented:
Yeah that's fair, NRF is special case due to the SoftDevices. TARGET_LPC810  and TARGET_KL05Z is not supported in mbed 5. We may also make an exception for Cortex A platform, but we need to document it well (your comment above gives nice explanation).

mprse commented:
@pan Can you check if current form is acceptable for NORDIC target?

pan- commented on targets/TARGET_NORDIC/TARGET_MCU_NRF51822/device/TOOLCHAIN_GCC_ARM/TARGET_MCU_NORDIC_32K/NRF51822.ld:
```diff
@@ -130,7 +130,7 @@ SECTIONS
 		__HeapBase = .;
 		*(.heap*)
-		. = ORIGIN(RAM) + LENGTH(RAM) - Stack_Size;
+		. += (ORIGIN(RAM) + LENGTH(RAM) - Stack_Size) - .;
```
Setting the pointer is much more clear. 
Side note: does that arithmetic operation use signed or unsigned integer math as this may lead to incorrect values if unsigned integer is used and the left operand less than the right operand.

pan- commented on targets/TARGET_NORDIC/TARGET_MCU_NRF51822/device/TOOLCHAIN_IAR/TARGET_MCU_NORDIC_32K/nRF51822_QFAA.icf:
```diff
@@ -9,8 +9,8 @@ define symbol __ICFEDIT_region_ROM_end__     = 0x0003FFFF;
 /*-Sizes-*/
-/*Heap 1/4 of ram and stack 1/8*/
-define symbol __ICFEDIT_size_cstack__   = 0xc00;
+/*Heap 1/4 of ram and stack 1024 B*/
```
2048 B

pan- commented on targets/TARGET_NORDIC/TARGET_MCU_NRF51822/device/TOOLCHAIN_GCC_ARM/TARGET_MCU_NRF51_16K_S110/NRF51822.ld:
```diff
@@ -130,7 +130,7 @@ SECTIONS
 		__HeapBase = .;
 		*(.heap*)
-		. = ORIGIN(RAM) + LENGTH(RAM) - Stack_Size;
+		. += (ORIGIN(RAM) + LENGTH(RAM) - Stack_Size) - .;
```
Please set the pointer.

pan- commented on targets/TARGET_NORDIC/TARGET_MCU_NRF51822/device/TOOLCHAIN_GCC_ARM/TARGET_MCU_NRF51_16K_S130/NRF51822.ld:
```diff
@@ -130,7 +130,7 @@ SECTIONS
 		__HeapBase = .;
 		*(.heap*)
-		. = ORIGIN(RAM) + LENGTH(RAM) - Stack_Size;
+		. += (ORIGIN(RAM) + LENGTH(RAM) - Stack_Size) - .;
```
Please set the pointer.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF51/TARGET_MCU_NRF51822_UNIFIED/device/TOOLCHAIN_GCC_ARM/TARGET_MCU_NORDIC_32K/NRF51822.ld:
```diff
@@ -144,7 +144,7 @@ SECTIONS
 		__HeapBase = .;
 		*(.heap*)
-		. = ORIGIN(RAM) + LENGTH(RAM) - Stack_Size;
+		. += (ORIGIN(RAM) + LENGTH(RAM) - Stack_Size) - .;
```
Please set the pointer.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF51/TARGET_MCU_NRF51822_UNIFIED/device/TOOLCHAIN_GCC_ARM/TARGET_MCU_NRF51_16K_S110/NRF51822.ld:
```diff
@@ -130,7 +130,7 @@ SECTIONS
 		__HeapBase = .;
 		*(.heap*)
-		. = ORIGIN(RAM) + LENGTH(RAM) - Stack_Size;
+		. += (ORIGIN(RAM) + LENGTH(RAM) - Stack_Size) - .;
```
Please set the pointer.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF51/TARGET_MCU_NRF51822_UNIFIED/device/TOOLCHAIN_GCC_ARM/TARGET_MCU_NRF51_16K_S130/NRF51822.ld:
```diff
@@ -142,7 +142,7 @@ SECTIONS
 		__HeapBase = .;
 		*(.heap*)
-		. = ORIGIN(RAM) + LENGTH(RAM) - Stack_Size;
+		. += (ORIGIN(RAM) + LENGTH(RAM) - Stack_Size) - .;
```
Please set the pointer.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52832/device/TOOLCHAIN_GCC_ARM/NRF52832.ld:
```diff
@@ -230,8 +232,8 @@ SECTIONS
-        ASSERT(. <= (ORIGIN(RAM) + LENGTH(RAM) - 0x800), "heap region overflowed into stack");
-        . += (ORIGIN(RAM) + LENGTH(RAM) - 0x800) - .;
+        ASSERT(. <= (ORIGIN(RAM) + LENGTH(RAM) - STACK_SIZE), "heap region overflowed into stack");
+        . += (ORIGIN(RAM) + LENGTH(RAM) - STACK_SIZE) - .;
```
Please set the pointer.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/TARGET_MCU_NRF52840/device/TOOLCHAIN_GCC_ARM/NRF52840.ld:
```diff
@@ -229,8 +231,8 @@ SECTIONS
-        ASSERT(. <= (ORIGIN(RAM) + LENGTH(RAM) - 0x800), "heap region overflowed into stack");
-        . += (ORIGIN(RAM) + LENGTH(RAM) - 0x800) - .;
+        ASSERT(. <= (ORIGIN(RAM) + LENGTH(RAM) - STACK_SIZE), "heap region overflowed into stack");
+        . += (ORIGIN(RAM) + LENGTH(RAM) - STACK_SIZE) - .;
```
Please set the pointer.

mprse commented:
Regarding to the following comment https://github.com/ARMmbed/mbed-os/pull/7079#issuecomment-400333649 :
> Good point, @mprse can you make sure we are not limiting stack to 1K on RTOS-less Mbed.

It looks like I need to define ISR/main stack size conditionally and set it to 1 KB (with some exceptions) for builds with RTOS and to at least 4 KB for builds without RTOS.

@SeppoTakalo @deepikabhavnani @0xc0170 Do you know maybe which symbol can be used to determine if build is with or without RTOS?

0xc0170 commented:
> It looks like I need to define ISR/main stack size conditionally and set it to 1 KB (with some exceptions) for builds with RTOS and to at least 4 KB for builds without RTOS.

👍  I was going to ask about it, thanks to Steve for that one.

> @SeppoTakalo @deepikabhavnani @0xc0170 Do you know maybe what symbol can be used to determine if build is with or without RTOS?

This symbol `MBED_CONF_RTOS_PRESENT`

pan- reviewed (dismissed):
Thanks for the changes!

0xc0170 commented:
@mprse Please provide an update here - I am after: what should be the sizes for rtos and nonrtos targets. 

There are few PR adding new targets currently opened - should they fix it within their PR (if we know the sizes we will advise them to change asap) ?

mprse commented:
Stack should be defined as described in the PR description (with some exceptions for ISR stack - Cortex A and NRF):

> 
> - ISR stack: exactly 1 kB,
> - Main thread stack: exactly 4 kB,
> - Default thread stack: exactly 4 kB,
> - RTOS-less thread stack: at least 4 kB

The problem is that we have same ISR/main stack definitions used for rtos and nonrtos builds. I'm currently investigating how to specify ISR/main stack based on `MBED_CONF_RTOS_PRESENT` symbol provided in `mbed_config.h`. I got some findings and I will add some proposition to this PR tomorrow (at least for GCC_ARM and ARM compilers). 

The main problem is how to make `MBED_CONF_RTOS_PRESENT` visible in the liker scripts. 
@theotherjimmy @deepikabhavnani @SeppoTakalo Do we have some solutions for such case?

deepikabhavnani commented:
> The main problem is how to make MBED_CONF_RTOS_PRESENT visible in the liker scripts.

Flags / defines should be available for use in linker scripts from tools point of view. See MBED_APP_START /  MBED_APP_SIZE macros usage as they are available in many linker scripts. 

Way to check defines will vary based on toolchain
https://github.com/ARMmbed/mbed-os/blob/8998a55bc78abb8ee69635c5b25ac46813d20a07/targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32GG/device/TARGET_1024K/TOOLCHAIN_ARM_STD/efm32gg.sct#L6

https://github.com/ARMmbed/mbed-os/blob/a0b624b62ebed56601548efea2822e27ecbba5b2/targets/TARGET_NXP/TARGET_LPC176X/device/TOOLCHAIN_IAR/LPC17xx.icf#L1

https://github.com/ARMmbed/mbed-os/blob/8998a55bc78abb8ee69635c5b25ac46813d20a07/targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32GG/device/TARGET_1024K/TOOLCHAIN_GCC_ARM/efm32gg.ld#L22

mprse commented:
Thanks @deepikabhavnani for the provided examples, but it looks like macros from `mbed_config.h` are not available in linker scripts.
In case of `GCC_ARM` compiler `MBED_CONF_RTOS_PRESENT` macro can be accessed via startup asm file. Then based on `MBED_CONF_RTOS_PRESENT` stack size can be defined in asm file and will be accessible in linker script.
In case of `ARM` compiler scatter file can be passed through preprocessor  and then `mbd_config.h` header can be simply included and `MBED_CONF_RTOS_PRESENT` symbol can be accessed.
I didn't found workaround for IAR.

deepikabhavnani commented:
>  but it looks like macros from mbed_config.h are not available in linker scripts.

@theotherjimmy - Is it possible to add config flags or macros to linker flags, or for this case does it make sense to detect rtos present and add `MBED_CONF_RTOS_PRESENT` flag to `self.flags["ld"]`?

mprse commented:
I have pushed commit with the proposition how we can distinguish builds with and without RTOS (for `GCC_ARM` and `ARM` compilers) - https://github.com/ARMmbed/mbed-os/pull/7238/commits/ff6c429d47f6836ab5446893f84b0007f16eada5.

I didn't found solution for `IAR` and I'm not sure if this solution can be used for all targets. It would be easier/better if we could use `mbed_config.h` symbols in linker scripts straightforward and it looks such such feature is comming (PR https://github.com/ARMmbed/mbed-os/pull/7061).

cmonr commented:
@mprse Looks like this will need a rebase.

@bulislaw @SeppoTakalo @deepikabhavnani @pan- @ARMmbed/mbed-os-core  Mind taking another look?

deepikabhavnani commented:
> I didn't found solution for IAR and I'm not sure if this solution can be used for all targets.

@cmonr - This looks work in progress and needs some solution for IAR, Linked PR #7061 might help.

cmonr commented:
@deepikabhavnani Reference: https://github.com/ARMmbed/mbed-os/pull/7061#issuecomment-402790138

That PR isn't addressing anything related to linkers.

deepikabhavnani commented:
@cmonr  - Yes 7061 is for adding macros to ASM files for all three toolchains, and similar approach can be used for linker flags as well? It would be a question for @theotherjimmy - Is it possible to add config flags or macros to linker flags?

theotherjimmy commented:
@deepikabhavnani It is possible, but it does not use the C preprocessor.

cmonr commented:
@mprse @SeppoTakalo @bulislaw @donatieng After a brief offline discussion, I'm wondering where this requirement is coming from.

Historically, we've been fairly hands-off with partner code and their implementations. Aligning buffer sizes is something that I would expect to come in as a coordinated effort from the PE team (@screamerbg @maclobdell). What's the reason for adding this test? Do we have partner buy-in already?

I don't like that these tests already require exceptions to be made for particular families. Imo, tests should not have as few baked-in exceptions as possible.

> Thanks @deepikabhavnani for the provided examples, but it looks like macros from mbed_config.h are not available in linker scripts.

@theotherjimmy mentioned offline that this is because *.h preprocessor symbols are meant to be used with the compiler, not the linker. The two applications require different symbols.

It's _possible_ that something could be setup to use the config system to set a default size, with target-specific overrides. Would something like that work?

Btw, I'm still confused as to how https://github.com/ARMmbed/mbed-os/pull/7061 would have resolved the issue. (In case you weren't aware, you can actually check out PRs and work with them locally.)

@ARMmbed/mbed-os-maintainers FYI.

cmonr reviewed (dismissed):
See above comment.

mprse commented:
> @mprse Looks like this will need a rebase.

Yes I will have to go through all linker scripts again when we find a proper way to define ISR stack size for builds with and without RTOS.

>@theotherjimmy - Is it possible to add config flags or macros to linker flags?
@deepikabhavnani It is possible, but it does not use the C preprocessor.

Do you plan to add such feature? 

> After a brief offline discussion, I'm wondering where this requirement is coming from.

Information about this task can be found here: https://jira.arm.com/browse/IOTHAL-18

> Btw, I'm still confused as to how #7061 would have resolved the issue. (In case you weren't aware, you can actually check out PRs and work with them locally.)

Thanks, I didn't check this since I was out of office. I just found #7061 (which was under development) and incorrectly assumed that this patch can potentially solve my problem.

bulislaw commented:
@cmonr stack sizes have little to do with partners and a lot to do with usability and consistency. There's no need for most of partners to tweak default ISR stack size, in the unlikely chance someone needs to touch it that will be the app dev. Lots of boards waste memory setting that to 4K or 16K either by mistake or not understanding how it works. That's the first, more externally visible change. In the next stage we need to think about unifying linker scripts using templates that platforms can tweak/extend, but by default they don't need to do anything if they don't have special circumstances.

theotherjimmy commented:
> Do you plan to add such feature?

Don't assume that I have not added the feature. I'll give you a hit: Managed and Unmanaged bootloader mode already pass in application start and size to the linker :wink:

theotherjimmy commented:
@bulislaw We almost have the info we need in the tools tools to generate linker scripts on the fly. We're missing ram/rom for targets that don't have a CMSIS-Pack.

mprse commented:
> Don't assume that I have not added the feature. I'll give you a hit: Managed and Unmanaged bootloader mode already pass in application start and size to the linker 😉

I started to search about "Managed and Unmanaged bootloader" and I found some information [here](https://docs.mbed.com/docs/mbed-os-handbook/en/latest/advanced/bootloader/) about application start and size: "The build system does this automatically by defining the symbols `MBED_APP_START` and `MBED_APP_SIZE` in the linker script (.sct, .ld or .icf).". But I'm not sure how can I make usage of this. I was thinking that maybe these symbols are defined only with RTOS builds, but I checked on `K64F/GCC_ARM` and it looks like `MBED_APP_START` is undefined in both cases (RTOS and nonRTOS build).

mprse commented:
@theotherjimmy  Now I see your point. `MBED_APP_START` and `MBED_APP_SIZE` are already passed to the linker scripts and I need to do it in the similar way with my `MBED_CONF_RTOS_PRESENT` symbol. This means that I need to provide some changes in the python build scripts 😕.

I found some code related to managed/unmanaged bootloader, application start and size here:
https://github.com/ARMmbed/mbed-os/blob/69d8c0bac3de45563976a1f5b32b74dd224978f7/tools/config/__init__.py#L44-L55

But for the moment I don't know how to add here `MBED_CONF_RTOS_PRESENT` config symbol. Can you give me some hint?

mprse commented:
I found solution to pass `MBED_CONF_RTOS_PRESENT` symbol to the linker script. Probably it is not the optimal solution, but it works. @deepikabhavnani @theotherjimmy Can you review this [change](https://github.com/ARMmbed/mbed-os/pull/7238/commits/d23b09454b59dc7c95b760330b5e90110cf4c354)?

Since now I can use `MBED_CONF_RTOS_PRESENT` I also provided some example changes in linker scripts (for few Freescale boards) which can be found [here](https://github.com/ARMmbed/mbed-os/pull/7238/commits/e7906cc39beda3fd2f3e48c1c0893590da39cc58). It would be great if someone who has more to do with linker scripts validated this before I proceed with the next linker scripts. There is a lot of files to change, so I would like to have agreed version. @bulislaw @SeppoTakalo @deepikabhavnani @0xc0170 @cmonr can you take a look?

theotherjimmy reviewed (changes requested):
Do not pass in a specific libraries macro into the linker scripts.

Do not scan the config header file when you _already have the config data_. That's just asking for trouble.

theotherjimmy commented:
@mprse We have a design goal of treating all libraries equal in the tools. You're introducing a special case for some library called `rtos`. Not only does this break the design goal, it's also not directly related to changing stack sizes!

mprse commented:
> @mprse We have a design goal of treating all libraries equal in the tools. You're introducing a special case for some library called rtos. Not only does this break the design goal, it's also not directly related to changing stack sizes!

@theotherjimmy I was nearly sure that my proposition will be rejected, but please note that I'm not familiar with python build scripts, libraries and their designs. I used your hint to do it in the similar way as it is done with  "Managed and Unmanaged bootloader, application start and size". So how this should be done to not break the design goal?

mprse commented:
@theotherjimmy ping.

theotherjimmy commented:
@mprse The problem I have is that you're creating a very bad precedent: If I merge this PR, what can I say to any community member that wants their library treated specially by the tools? My hint was to point out that `target` is not a library, and already affects the build process. We tend to apply a bit more scrutiny to new target configuration, as it affects everyone.

mprse commented:
@theotherjimmy Thanks for clarification. 
@bulislaw @0xc0170 @cmonr 
The conclusion is that modification of build script in order to pass `MBED_CONF_RTOS_PRESENT` symbol to the linker script is not acceptable solution. 
Unfortunately at the moment I can't see other option how to define different ISR stack size for builds with and without RTOS. Any ideas? Maybe we should have 2 linker scripts (one for builds with RTOS and one without RTOS)?

> @bulislaw We almost have the info we need in the tools tools to generate linker scripts on the fly. We're missing ram/rom for targets that don't have a CMSIS-Pack.

Maybe if you plan to generate linker scripts on the fly in the future, then stack size unification should be considered then?

c1728p9 commented on targets/TARGET_ARM_FM/mbed_rtx.h:
```diff
@@ -25,4 +27,16 @@
+    extern uint32_t               Image$$ARM_LIB_STACK$$ZI$$Base[];
+    extern uint32_t               Image$$ARM_LIB_STACK$$ZI$$Length[];
+    #define ISR_STACK_START       ((unsigned char*)Image$$ARM_LIB_STACK$$ZI$$Base)
+    #define ISR_STACK_SIZE        ((uint32_t)Image$$ARM_LIB_STACK$$ZI$$Length)
```
It looks like this is the same for most targets. Could be moved to a common location like alongside these values:
https://github.com/ARMmbed/mbed-os/blob/be215a310d51c694a992e913f6757e96c397cefd/rtos/TARGET_CORTEX/mbed_boot.c#L238-L245

Also, you'll also need to update `HEAP_START` and `HEAP_SIZE` so there is no overlap between it and `ISR_STACK_START` and `ISR_STACK_SIZE`. You may also be able to remove `INITIAL_SP` in the process since 'ISR_STACK_START' should be specifying the same information but in a way that won't get out of sync with the linker script.

c1728p9 commented:
Hi @mprse below is my brain dump on this PR. Hopefully it is useful.

[mbed 5 layout](https://os.mbed.com/docs/v5.9/reference/memory.html)
- Fixed sized heap
- Fixed 4K main stack
- Fixed 1K interrupt stack

[mbed 2 layout](https://os.mbed.com/handbook/Memory-Model)
- Variable sized heap (shared with stack)
- Variable sized main stack (shared with heap)
- Interrupt stack is the main stack

Most (all?) ARMCC and GCC linker scripts adhere to the above layouts. Typically the ISR stack size is not specified so it uses the [default of 1K](https://github.com/ARMmbed/mbed-os/blob/be215a310d51c694a992e913f6757e96c397cefd/rtos/TARGET_CORTEX/mbed_boot.c#L249-L252).

IAR is problematic as it requires a fixed sized heap and stack to be specified by the linker script. It does not adhere to the layout mentioned above for mbed 2. Some possible options to align IAR with the rest of the toolchains:
- Pull in a newer version of IAR which supports variable sized heap. I think @deepikabhavnani was looking into this at one point
- Pass an RTOS define to the IAR linker script so stack size can be adjusted for 2 or 5 based on that. Note - the memory model would still not match mbed 2, but the stack could be sized more reasonably.
- Just use the 1K (ISR) stack size that mbed 5 expects in all cases

mprse commented:
@c1728p9 Thanks for shedding light on the topic. This is very helpful.

deepikabhavnani commented:
> Pull in a newer version of IAR which supports variable sized heap. I think @deepikabhavnani was looking into this at one point

We cannot pull in higher version of IAR compiler( 8 or higher) which supports variable size heap, as it supports C++14 only.

mprse commented:
> Just use the 1K (ISR) stack size that mbed 5 expects in all cases

We can't do that since for mbed 2 we need more space for main stack and for mbed 5 we need some exceptions for special cases like Nordic boards with SoftDevices or Cortex-A platforms.

mprse committed:
Define ISR_STACK_START and ISR_STACK_SIZE symbols in mbed_boot.c

For mbed 5 ISR stack size should be equal to 1 KB (with some exceptions).
Define ISR_STACK_START and ISR_STACK_SIZE symbols based on linker scripts symbols.

Additionally remove `INITIAL_SP` in the process since `ISR_STACK_START` is now specifying the same information but in a way that won't get out of sync with the linker script.

mprse committed:
TARGET_Freescale: ISR stack size unification

For boards which supports MBED 2 and MBED 5 configure ISR stack size as follows (GCC_ARM, ARM):
MBED 5: 1 KB
MBED 2: 4 KB

The exception is IAR since currently there is no way to define different stack sizes for builds with and without RTOS. In this case ISR stack is set to 1 KB which is consistent with MBED 5 builds. For builds without RTOS if larger ISR stack size is needed it has to be modified manually in ICF file.

BOARD       RELEASE    GCC_ARM   ARM    IAR
K20D50M     [2]        -         -      -
TEENSY3_1   [2]        -         -      -
KL05Z       [2]        -         -      -
KL25Z       [2, 5]     done      done   fixed isr stack 1KB
KL26Z       []         -         -      -
KL46Z       [2, 5]     done      done   fixed isr stack 1KB
K66F        [2, 5]     done      done   fixed isr stack 1KB
K82F        [2, 5]     done      done   fixed isr stack 1KB
KL27Z       [2]        -         -      -
KL43Z       [2, 5]     done      done   fixed isr stack 1KB
KL82Z       [2, 5]     done      done   fixed isr stack 1KB
KW24D       [2, 5]     done      done   fixed isr stack 1KB
KW41Z       [2, 5]     done      done   fixed isr stack 1KB
K22F        [2, 5]     done      done   fixed isr stack 1KB
K64F        [2, 5]     done      done   fixed isr stack 1KB

mprse commented:
Moved definition of `ISR_STACK_START` and `ISR_STACK_SIZE` from `mbed_rtx.h` files to the common location (`mbed_boot.c`) and removed `INITIAL_SP` in the process since `ISR_STACK_START` is now specifying the same information, but in a way that won't get out of sync with the linker script. These has been suggested by @c1728p9.

Current proposition does not pass RTOS define to the linker scripts as this idea has been rejected by @theotherjimmy. As a result it defines different ISR stack sizes for builds with(mbed 5) and without(mbed 2) RTOS only for `ARM` and `GCC_ARM` compilers. For `IAR` fixed 1K ISR stack size that mbed 5 expects is used. I assume that for builds without RTOS stack size will have to be changed manually if needed. Some exception will have to be added in cases mentioned earlier (Nordic boards with SoftDevices or Cortex-A platforms).

At the moment this PR provides stack size unification only for `Freescale` boards:

<table>
<tr><th>BOARD       <th>RELEASE    <th>GCC_ARM   <th>ARM    <th>IAR
<tr><td>K20D50M     <td>[2]       <td> -        <td> -      <td>-
<tr><td>TEENSY3_1   <td>[2]       <td> -        <td> -      <td>-
<tr><td>KL05Z       <td>[2]        <td>-         <td>-      <td>-
<tr><td>KL25Z       <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>KL26Z       <td>[]         <td>-         <td>-      <td>-
<tr><td>KL46Z       <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>K66F        <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>K82F        <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>KL27Z       <td>[2]        <td>-         <td>-      <td>-
<tr><td>KL43Z       <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>KL82Z       <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>KW24D       <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>KW41Z       <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>K22F        <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
<tr><td>K64F        <td>[2, 5]     <td>done      <td>done   <td>fixed isr stack 1KB
</table>

@bulislaw @c1728p9 @SeppoTakalo @deepikabhavnani @0xc0170 @cmonr Guys do you mind to take a look if this make sense and if the solution with fixed 1KB ISR stack size for `IAR` is acceptable?

c1728p9 reviewed (approved):
@mprse this looks good to me. Having a 1KB stack for IAR seems fine to me. The subset of users using IAR to build mbed 2 only is likely small, and even then  they can workaround it by tweaking this value themselves. It might be good to get a second opinion on this though.

mprse commented:
@c1728p9 Thanks for your opinion. As you suggested I will wait for the others before moving forward with this. The most important for me is to validate linker scripts if such changes are correct before I will change all in the same manner which is a lot of work.

SeppoTakalo commented:
I agree that Mbed2+IAR fixed 1kB is acceptable as there are easy workarounds. Otherwise it might be too complex to fully unify these, as all linkers behave a bit differently.

cmonr commented:
@mprse Making sure I understand corectly. The scope of this PR is now just convering Freescale targets?
Would you mind updating the PR title if that's the case?

mprse commented:
@cmonr No, this PR is about stack size unification for all targets. This is a lot of changes in many linker script, so I created a sample for verification/review before further processing.

0xc0170 commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/device/TOOLCHAIN_IAR/MK64FN1M0xxx12.icf:
```diff
@@ -49,9 +49,13 @@
-define symbol __stack_size__=0x8000;
-define symbol __heap_size__=0x10000;
+/* Stack and Heap sizes */
+if (isdefinedsymbol(MBED_CONF_RTOS_PRESENT)) {
```
Based on the latest comments, this does not work and will be just 0x400 as for other targets like KL25z/46z etc ?

mprse replied:
I missed this file while updates. Thanks!

Yes, this should be like in other freescale boards:
define symbol __stack_size__=0x400;

0xc0170 commented:
It was mentioned above but would like to have definite answer by this PR - as this unifies stack sizes . WE often got asked - can this be configured ? Not editing the linker script file but rather config. Looking at the changes here , the values are still hardcoded

There are couple of mentions above about preprocess linker files. This was turned down,  I recall discussions about generating linker script files. So it might be out of the scope for this PR or can be part of it (if we allow to preprocess linker script file with just one -D and that is stack size).

mprse commented:
Thanks guys for the review. @0xc0170 question is not answered about generating linker script files/configured stack size, but since there are no objections to this solution I'll continue working on this.

mprse committed:
TARGET_NXP: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 - 0 boards:
----------------

 MBED 2, 5 - 9 boards:
----------------
FF_LPC546XX
ARCH_PRO
UBLOX_C027
LPC546XX
LPC4088_DM
MIMXRT1050_EVK
LPC4088
LPC54114
LPC1768

 MBED 2 - 15 boards (skipped):
----------------
LPC1549
OC_MBUINO
LPC4337
ARCH_GPRS
LPC11U24
LPC1347
LPC824
LPC11U68
LPC11U35_401
LPC11U37H_401
LPC812
LPC1114
MICRONFCBOARD
LPC11U35_501
SSCI824

mprse committed:
TARGET_Nordic: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).
NOTE: Exception for Nordic board. Since nordic boards are using SoftDevice ISR Stack must be set to 2 KB.

 MBED 5 : 10 boards
----------------
MTB_LAIRD_BL600
MTB_LAIRD_BL652
NRF52_DK
NRF52840_DK
DELTA_DFBM_NQ620
UBLOX_EVK_NINA_B1
VBLUNO52
MTB_UBLOX_NINA_B1
RBLAB_BLENANO2
UBLOX_EVA_NINA
 MBED 2, 5 : 3 boards
----------------
NRF51_DK
TY51822R3
NRF51_DONGLE
 MBED 2 : 12 boards (skipped)
----------------
RBLAB_NRF51822
RBLAB_BLENANO
ARCH_BLE
VBLUNO51
MTM_MTCONNECT04S
NRF51822
SEEED_TINY_BLE
NRF51_DONGLE_LEGACY
DELTA_DFCM_NNN40
NRF51_MICROBIT_B
WALLBOT_BLE
HRM1017

mprse committed:
TARGET_Analog_Devices: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 : 2 boards
----------------
EV_COG_AD4050LZ
EV_COG_AD3029LZ

 MBED 2, 5 : 0 boards
----------------

 MBED 2 : 0 boards (skipped)
----------------

mprse committed:
TARGET_Arm_SSG: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 : 0 boards
----------------

 MBED 2, 5 : 2 boards
----------------
ARM_CM3DS_MPS2
ARM_BEETLE_SOC

 MBED 2 : 6 boards (skipped)
----------------
ARM_MPS2_M0
ARM_MPS2_M3
ARM_MPS2_M4
ARM_MPS2_M7
ARM_MPS2_M0P
ARM_IOTSS_BEID

mprse committed:
TARGET_ONSEMI: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 : 0 boards
----------------

 MBED 2, 5 : 1 boards
----------------
NCS36510

 MBED 2 : 0 boards (skipped)
----------------

mprse committed:
TARGET_Realtek: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 : 1 boards
----------------
REALTEK_RTL8195AM

 MBED 2, 5 : 0 boards
----------------

 MBED 2 : 0 boards (skipped)
----------------

mprse committed:
TARGET_TOSHIBA: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 : 3 boards
----------------
TMPM3H6
TMPM46B
TMPM066
 MBED 2, 5 : 0 boards
----------------
 MBED 2 : 0 boards (skipped)
----------------

mprse commented:
Status of stack unification per Vendor.

Target | Status
-- | --
TARGET_Analog_Devices | Done
TARGET_ARM_FM | https://github.com/ARMmbed/mbed-os/pull/7706
TARGET_ARM_SSG | Done
TARGET_Atmel | Done (Nothing to do - no MBED 5 boards)
TARGET_Freescale | Done
TARGET_Maxim |  
TARGET_NORDIC | Done
TARGET_NUVOTON |  
TARGET_NXP | Done
TARGET_ONSEMI | Done
TARGET_Realtek | Done
TARGET_RENESAS | Cortex-A boards - skipped 
TARGET_Silicon_Labs | Done
TARGET_STM |  WIP
TARGET_TOSHIBA | Done
TARGET_ublox | Done (Nothing to do - no MBED 5 boards)
TARGET_WICED | Done (Nothing to do - bins only (?) )
TARGET_WIZNET | Done

mprse committed:
TARGET_WIZNET: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

 MBED 5 : 0 boards
----------------

 MBED 2, 5 : 3 boards
----------------
WIZWIKI_W7500ECO
WIZWIKI_W7500P
WIZWIKI_W7500

 MBED 2 : 0 boards (skipped)
----------------

mprse committed:
TARGET_Silicon_Labs: ISR stack size unification

Unify ISR stack size for targets which support MBED 5 (or MBED 2 and MBED 5).

c1728p9 commented:
@mprse one thing to be aware of is you may need special handling for backward compatibility with [mbed-rtos](https://os.mbed.com/users/mbed_official/code/mbed-rtos/file/5713cbbdb706/rtx/TARGET_CORTEX_M/RTX_CM_lib.h/).  I think it uses the value INITIAL_SP and allocates memory based on a single region layout by default.

0xc0170 commented:
@mprse Is this work inteded for 5.10 ?

donatieng commented:
After discussing with @c1728p9 we decided to hold this PR till 5.10 is out and get it merged to master just afterwards as it impacts a wide number of boards not available in CI. This would allow us to mitigate any side effects and provide any required fixes before 5.11.

cmonr commented:
@donatieng In that case, do y'all want a feature branch for this work?

donatieng commented:
@cmonr I think work is mostly complete so a feature branch should not be required, what do you think @c1728p9 @mprse?

mprse commented:
> @mprse one thing to be aware of is you may need special handling for backward compatibility with mbed-rtos. I think it uses the value `INITIAL_SP` and allocates memory based on a single region layout by default.

So it looks like `INITIAL_SP` can not be removed permanently. It should  be defined as `ISR_STACK_START`.

mprse commented:
Rebased.

mprse committed:
TARGET_STM: ISR stack size unification - WIP

The following Mbed 5 boards were processed:
NUCLEO_F091RC
NUCLEO_F070RB
NUCLEO_F072RB

NUCLEO_F103RB

NUCLEO_F207ZG

DISCO_F303VC
NUCLEO_F303ZE
NUCLEO_F303RE

--- only boards marked with "-" of NUCLEO_F3 family were processed
UBLOX_C030_U201
MTB_UBLOX_ODIN_W2
MTB_MXCHIP_EMW3166
MBED_CONNECT_ODIN
UBLOX_C030_N211
UBLOX_EVK_ODIN_W2
USI_WM_BN_BM_22
UBLOX_C030_R410M
NUCLEO_F411RE
NUCLEO_F413ZH
DISCO_F407VG
- NUCLEO_F401RE
NUCLEO_F429ZI
B96B_F446VE
DISCO_F413ZH
NUCLEO_F439ZI
DISCO_F429ZI
- MTS_DRAGONFLY_F411RE
NUCLEO_F446ZE
DISCO_F469NI
NUCLEO_F446RE
NUCLEO_F410RB
MTS_MDOT_F411RE
WIO_3G
STEVAL_3DP001V1
- MTB_MTS_DRAGONFLY
NUCLEO_F412ZG

jeromecoutant commented:
Hi

Maybe we can push a little bit this PR as more and more PR on the same topic comes ?
#8225 #8252 #7909 #7289 ...

Thx

jeromecoutant commented:
Maybe should we implement proposition from #8252:

if (!isdefinedsymbol(MBED_APP_HEAP_SIZE)) {
define symbol size_heap = <default size>;
} else {
define symbol size_heap = MBED_APP_HEAP_SIZE;
}

mprse commented:
I will have to mainly review all files again when framework for configuring boot stack size will go in (PR https://github.com/ARMmbed/mbed-os/pull/8039).

cmonr commented:
@mprse Is this still in progress? #8039 has since been merged

0xc0170 commented:
@mprse Let us know the status. We shall close it and reopen with an update? Or is there expecting completion anytime soon?

mprse commented:
> @mprse Let us know the status. We shall close it and reopen with an update? Or is there expecting completion anytime soon?

I'm planning to continue this one when I'm done with SPI testing. Since we have new Framework added by @c1728p9 (PR https://github.com/ARMmbed/mbed-os/pull/8039) I will have to go through all linker scripts again, so I think we should close this one and I'll open new one instead later.

0xc0170 commented:
Sounds good to me (reference this previous work if needed any history). I'll close this one
